java/test/src/main/java/test/Ice/ami/AllTests.java

```diff
@@ -966,12 +970,11 @@ public static void allTests(test.TestHelper helper, boolean collocated) {
                         maxQueue *= 2;
                         done = false;
                     }
-                    for (InvocationFuture<Void> q : results) {
-                        q.join();
-                    }
+                    CompletableFuture.allOf(results.toArray(new CompletableFuture[0])).join();
+
+                    // Wait until the connection is closed.
+                    p.ice_getCachedConnection().close();
                 }
-                // Wait until the connection is closed.
-                p.ice_getCachedConnection().close();
             }
             out.println("ok");
```

Comment:
I think this would fix #3049

Reply:
I don't think this change is correct.

In the original code/test, the while loop sends a bunch of opWithPayloadAsync and closeConnectionAsync invocations to the server. The expectation is all these calls will succeed; for opWithPayload received by the server after the server sent CloseConnection to the client, they succeed because the client "retries" and opens a new connection.

Now, in the original code, the client never explicitly closes the connection in the while loop. That's the whole point of this test. The close outside the while loop (at the very end) is to take care of the connection established by an opWithPayload sent after a CloseConnection message was received from the server. You should not change this logic.

Reply:
Check the C++ and C# versions, they do the same.  If we don't wait for the close the setup when we have to retry is not correct.

At the point we close we have awaited all outgoing requests.